plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java

```diff
@@ -187,6 +192,16 @@ void executeSingleUpload(final S3BlobStore blobStore,
         if (blobStore.serverSideEncryption()) {
             md.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
         }
+        try {
+	    int readLimit = toIntExact(blobSize);
+	    input.mark(readLimit);
+	    byte[] md5 = Md5Utils.computeMD5Hash(input);
```

Comment:
This is not something we can do so easily I'm afraid. The reason is that we allow for setting a rate limit on this input stream. So this code would literally halve the throughput in a rate limited scenario.